data/src/main/java/com/github/dodobest/data/hilt/UpbitDataModule.kt

```diff
@@ -0,0 +1,57 @@
+package com.github.dodobest.data.hilt
+
+import com.github.dodobest.data.Constant
+import com.github.dodobest.data.data.UpbitAPI
+import com.github.dodobest.data.data.UpbitRemoteDataSource
+import com.github.dodobest.data.repository.UpbitRepositoryImpl
+import com.github.dodobest.domain.UpbitRepository
+import dagger.Module
+import dagger.Provides
+import dagger.hilt.InstallIn
+import dagger.hilt.components.SingletonComponent
+import hu.akarnokd.rxjava3.retrofit.RxJava3CallAdapterFactory
+import okhttp3.OkHttpClient
+import retrofit2.Retrofit
+import retrofit2.converter.gson.GsonConverterFactory
+import java.util.concurrent.TimeUnit
+
+@Module
+@InstallIn(SingletonComponent::class)
+object UpbitDataModule {
+    @Provides
+    fun provideUpbitRepository(
```

Comment:
```suggestion
    @Provides
    @Singleton
    fun provideUpbitRepository(
```
싱글턴 인스턴스를 가지게 하려면, Singleton 어노테이션을 붙여주세요. (이하 관련 내용 동일)

Reply:
[79d14d3](https://github.com/DoTheBestMayB/UpbitAPI/pull/26/commits/79d14d3a6d9e2a564aefaa6183e068cb557bf5cf)

`@InstallIn()`는 Hilt가 어디에 주입할지 결정해서 생성되는 객체의 LifeCycle을 결정하는 어노테이션입니다.
`ApplicationComponent`는 안드로이드 모듈에만 적용할 수 있는데, 안드로이드가 아닌 모듈에 `SingletonComponent`를 사용하면 됩니다.
ApplicationComponent`는 Activity와 LifeCycle이 같지만, `SingletonComponent`은 Application의 LifeCycle과 같습니다.

`@Singleton`은 DCL(Double Check Locking) 방식을 사용해서 객체를 이미 생성했다면 생성된 객체를 제공하라는 싱글턴 적용 어노테이션입니다.

https://stackoverflow.com/a/65012371/11722881
https://developer.android.com/training/dependency-injection/hilt-android#component-lifetimes

Reply:
참고로 ApplicationComponent가 SingletonComponent로 변경되었어요.
즉 둘 다 같은 의미에요.

둘다 Singleton이 붙어서 햇갈릴 수 있지만, 핵심은 `@InstallIn`은 Hilt가 의존성을 주입하는 시점을 결정하는 것이고, `@Singleton`은 객체를 싱글턴으로 만드는 것을 의미합니다.

Reply:
UpbitRepository를 Singleton으로 만들지 않으면, 해당 repository를 주입 받은 viewModel, fragment 등이 동시에 API를 요청할 경우, [API 요청 수 제한](https://github.com/DoTheBestMayB/UpbitAPI/issues/25)을 지키는 로직을 적용하기 어려울 것 같아요.
ex) 코인 시세를 보여주는 화면(activity)에서 특정 버튼을 누르면 팝업 창(fragment)을 보여주고, 여기서 사용자가 조회하고 싶은 코인 이름을 입력할 수 있도록 할 때, activity와 fragment가 각각 repository를 가진다면, 두 repository가 요청하는 API의 우선순위를 처리하는 클래스를 추가로 만들어야 할 것 같아요.

그래서 `UpbitRepositoryModule`을 Singleton으로 만들어야 한다고 생각했는데, 어떤가요?

Reply:
> UpbitRepository를 Singleton으로 만들지 않으면, 해당 repository를 주입 받은 viewModel, fragment 등이 동시에 API를 요청할 경우, https://github.com/DoTheBestMayB/UpbitAPI/issues/25을 지키는 로직을 적용하기 어려울 것 같아요.

싱글턴과 API 요청 수하고는 관련이 없어요.
왜냐하면 안드로이드는 화면 단위이기 때문에, 같은 API를 동시에 요청할 수가 없어요.
그리고 원래 public api에는 DDOS를 고려하여 api 요청 수 제한이 있어요. 만약 api 요청수 제한을 없애고 싶으시면 비용을 부담하시면 될거에요.

Repository는 일반적으로 자주 사용이 되기도 하고, 매번 API를 호출할 때마다 객체를 생성하는 비용을 줄여서 약간의 속도 측면에서 이점을 얻기 위해서 싱글턴으로 만들곤 합니다.

Reply:
[058c241](https://github.com/DoTheBestMayB/UpbitAPI/pull/26/commits/058c241423e72846562328ed4b6912b9b281446f)